File: Stacks & Queues/SortStack.java
// Cracking the Coding Interview - pp. 98 - q 3.5

import java.util.*;

public class SortStack {

    // Create a stack
    Stack<Integer> stack;

    SortStack() {
        stack = new Stack<Integer>();
    }

    void push(int newVal) {

        // OBJECTIVE: Insert newVal to stack in correct order

        // If stack is empty, add newVal to stack and exit function
        if (stack.isEmpty()) {
            stack.push(newVal);
            return;
        }

        // If stack isn't empty, create a temporary stack and pop all elements from stack to it until finding a perfect place to add newVal
        Stack<Integer> tmp = new Stack<Integer>();

        // Iterate stack
        while (!stack.isEmpty()) {

            // Pop element from stack
            int curHead = stack.peek();

            // If curHead is smaller than newVal, pop it from stack
            if (curHead < newVal) {
                tmp.push(stack.pop());
            }
            else {
                break;
            }
        }

        // Add newVal to stack
        stack.push(newVal);

        // Iterate tmp and add all elements back to stack
        while (!tmp.isEmpty()) {
            stack.push(tmp.pop());
        }
    }

    int pop() {

        // OBJECTIVE: Pop and return head element from stack
        return stack.pop();
    }

    int peek() {
        
        // OBJECTIVE: Return value of head element from stack
        return stack.peek();
    }

    void printStack() {

        // OBJECTIVE: Print stack

        for (int elem: stack) {
            System.out.printf("%d ", elem);
        }
    }
}

File: Stacks & Queues/Sort Stack.py
# Cracking the Coding Interview - pp. 98 - q 3.5

from stack import Stack

class SortStack:

    def __init__(self) -> None:

        # Initialize stack class
        self.stack = Stack()

    def push(self, newVal):

        # OBJECTIVE: Add new value to stack in ascending order

        # If stack is empty, add newVal to stack and exit function
        if self.stack.isEmpty():
            self.stack.push(newVal)
            return

        # If stack isn't empty, pop every element from stack until newVal is added in correct order
        tmp = Stack()

        # Iterate stack
        while not self.stack.isEmpty():
            
            # Pop element from stack
            curHead = self.stack.peek()

            # If popped is smaller than 
            if curHead < newVal:
                tmp.push(self.stack.pop())
            else:
                break

        # Add newVal to stack
        self.stack.push(newVal)

        # Add all elements from tmp back to stack
        while not tmp.isEmpty():
            self.stack.push(tmp.pop())

    def pop(self):

        # OBJECTIVE: Remove head element from stack and return it
        return self.stack.pop()

    def peek(self):

        # OBJECTIVE: Return value of head element from stack
        return self.stack.peek()

    def isEmpty(self):

        # OBJECTIVE: Return true/false if stack is empty
        return self.stack.isEmpty()

    def printStack(self):

        # OBJECTIVE: Print stack
        self.stack.printStack()

def main():

    # Initialize class
    sortedStack = SortStack()
    sortedStack.push(1)
    sortedStack.push(21)
    sortedStack.push(153)
    sortedStack.push(1123)
    sortedStack.push(0)

    sortedStack.printStack()

    sortedStack.pop()
    sortedStack.printStack()

main()